File: src/main/java/com/avast/syringe/config/internal/Injection.java
package com.avast.syringe.config.internal;

import com.avast.syringe.config.ConfigException;

import java.lang.reflect.Array;
import java.util.*;

/**
 * Performs the actual injection of a {@link Property} into an injectable property
 * of a given instance.
 */
public class Injection {

    public interface ContextualPropertyResolver {
        Object getProperty(InjectableProperty property, Value configValue) throws Exception;
    }

    public static final ContextualPropertyResolver DEFAULT_CONTEXTUAL_PROPERTY_RESOLVER = new ContextualPropertyResolver() {
        @Override
        public Object getProperty(InjectableProperty property, Value configValue) throws Exception {
            throw new NoSuchFieldException(property.getName());
        }
    };

    private final InjectableProperty property;

    public Injection(InjectableProperty property) {
        this.property = property;
    }

    /**
     * Takes the property from the {@code properties} map, converts it to the
     * correct value and sets it to the {@code instance}'s property.
     */
    public void apply(Object instance, Map<String, Property> properties, ContextualPropertyResolver resolver)
            throws Exception {
        if (property.isContextual() || property.isReference()) {
            applyContextual(instance, properties, resolver);
        } else {
            applyNonContextual(instance, properties, resolver);
        }
    }

    private void applyContextual(Object instance, Map<String, Property> properties, ContextualPropertyResolver resolver) throws Exception {
        Value configValue = getConfigValue(properties);
        Object propVal;
        try {
            propVal = resolver.getProperty(property, configValue);
        } catch (NoSuchFieldException e) {
            throwNotConfigured(e);
            return;
        }
        property.setValue(instance, propVal);
    }

    private void throwNotConfigured(NoSuchFieldException e) throws NoSuchFieldException {
        if (!property.isOptional()) {
            String message = String.format("Required property %s not configured", property.getName());
            throw e == null ? new NoSuchFieldException(message) : e;
        }
    }

    private Value getConfigValue(Map<String, Property> properties) throws Exception {
        Property configProp = properties.get(property.getName());
        Value configValue;
        if (configProp == null) {
            configValue = null;
        } else {
            try {
                configValue = configProp.getValues().get(0);
            } catch (Exception e) {
                throw e;
            }
        }
        return configValue;
    }

    public InjectableProperty getProperty() {
        return property;
    }

    private void applyNonContextual(Object instance, Map<String, Property> properties,
                                    ContextualPropertyResolver resolver) throws Exception {
        String propName = property.getName();
        Property prop = properties.get(propName);

        if (prop == null) {

            if (resolver != null) {
                applyContextual(instance, properties, resolver);
                return;
            } else {
                throwNotConfigured(null);
                return;
            }

        }

        if (property.isArray()) {
            injectArray(instance, prop, resolver);
        } else if (property.isCollection()) {
            Class<?> type = property.getType();
            if (List.class.isAssignableFrom(type)) {
                injectCollection(instance, prop, ArrayList.class, resolver);
            } else if (Set.class.isAssignableFrom(type)) {
                injectCollection(instance, prop, HashSet.class, resolver);
            } else {
                injectCollection(instance, prop, null, resolver);
            }
        } else if (property.isMap()) {
            injectMap(instance, prop, resolver);
        } else {
            injectSimple(instance, prop, resolver);
        }
    }

    private void injectSimple(Object instance, Property prop, ContextualPropertyResolver resolver) throws Exception {
        List<Value> values = prop.getValues();
        if (values.size() > 1) {
            String message = String.format("Property %s takes a single value, but was configured with %s", prop.getName(), prop.getValues());
            throw new ConfigException(prop, instance.getClass(), message, null);
        }

        Value value = prop.getValues().get(0);
        Object val;
        String refType = value.getRefType();
        if (refType != null) {
            val = resolver.getProperty(property, value);
        } else {
            val = TypeConversion.convert(value.getValue(), property.getType());
        }

        property.setValue(instance, val);
    }

    private void injectArray(Object instance, Property prop, ContextualPropertyResolver resolver) throws Exception {
        Class<?> componentType = property.getArrayOrCollectionComponentType();
        if (componentType.isArray()) {
            String message = String.format("Multi-dimensional array property %s is not supported. Only single-dimensional arrays are supported",
                    property.getName());
            throw new ConfigException(prop, instance.getClass(), message, null);
        }

        Object array = property.getValue(instance);
        if (array == null) {
            array = Array.newInstance(componentType, prop.getValues().size());
            property.setValue(instance, array);
        }

        for (int i = 0; i < prop.getValues().size(); i++) {
            Value value = prop.getValues().get(i);
            Object element;
            String refType = value.getRefType();
            if (refType != null) {
                element = resolver.getProperty(property, value);
            } else {
                element = TypeConversion.convert(value.getValue(), componentType);
            }
            property.setArrayElement(instance, i, element);
        }
    }

    @SuppressWarnings({"unchecked", "rawtypes"})
    private void injectCollection(Object instance, Property prop, Class<? extends Collection> defaultCollectionClass,
                                  ContextualPropertyResolver resolver) throws Exception {
        Collection<Object> collection = (Collection<Object>) property.getValue(instance);
        if (collection == null) {
            if (defaultCollectionClass == null) {
                String message = String.format("Cannot infer a collection implementation for property %s. Initialize it with a concrete empty collection",
                        property.getName());
                throw new ConfigException(prop, instance.getClass(), message, null);
            }
            collection = ConfigInjector.createInstance(defaultCollectionClass);
            property.setValue(instance, collection);
        }

        Class<?> componentType = property.getArrayOrCollectionComponentType();
        for (Value value : prop.getValues()) {
            Object element;
            String refType = value.getRefType();
            if (refType != null) {
                element = resolver.getProperty(property, value);
            } else {
                element = TypeConversion.convert(value.getValue(), componentType);
            }
            property.addCollectionElement(instance, element);
        }
    }

    @SuppressWarnings({"unchecked"})
    private void injectMap(Object instance, Property prop, ContextualPropertyResolver resolver) throws Exception {
        Map<Object, Object> map = (Map<Object, Object>) property.getValue(instance);
        if (map == null) {
            map = ConfigInjector.createInstance(HashMap.class);
            property.setValue(instance, map);
        }

        Class<?> keyType = property.getMapKeyType();
        Class<?> valueType = property.getMapValueType();
        for (Value value : prop.getValues()) {
            MapEntry entry = (MapEntry) value;
            Object mapKey = TypeConversion.convert(entry.getKey(), keyType);

            Object mapValue;
            String refType = value.getRefType();
            if (refType != null) {
                mapValue = resolver.getProperty(property, value);
            } else {
                mapValue = TypeConversion.convert(value.getValue(), valueType);
            }

            property.putMapEntry(instance, mapKey, mapValue);
        }
    }
    public static void inject(Object instance, String propertyName, Object value){
        InjectableProperty property = getInjectableProperty(instance.getClass(), propertyName);
        try {
            property.setValue(instance, value);
        } catch (Exception e) {
            throw new RuntimeException(e);
        }
    }
    private static InjectableProperty getInjectableProperty(Class cls, String propertyName){
        ConfigClassAnalyzer configClassAnalyzer = new ConfigClassAnalyzer(cls);
        for(InjectableProperty  property : configClassAnalyzer.getConfigProperties()){
            if(property.getName().equals(propertyName))
                return property;
        }
        return null;
    }
}

File: src/main/scala/com/avast/syringe/config/perspective/Injection.scala
package com.avast.syringe.config.perspective

import com.avast.syringe.config.internal.InjectableProperty

/**
 * User: slajchrt
 * Date: 6/5/12
 * Time: 3:40 PM
 */

object Injection {

  def apply(property: InjectableProperty,
            resolvers: List[PropertyResolver]): Injection = {

    property match {
      case _ if property.isArray => new ArrayInjection(property, resolvers)
      case _ if property.isCollection => new CollectionInjection(property, resolvers)
      case _ if property.isMap => new MapInjection(property, resolvers)
      case _ => new ScalarInjection(property, resolvers)
    }
  }

}

abstract class Injection(prop: InjectableProperty,
                         resolvers: List[PropertyResolver]) {

  def inject[T](instance: T, values: List[Any]): T

  protected def prepareValueForInjection(instance: Any): Any = {
    instance match {
      case builder: Builder[_] => builder.build
      case _ => instance
    }
  }

  protected def findAndSetContextualValue[T](instance: T, errorMessage: => String) {
    resolvers.find(_.hasPropertyValue(instance, prop)) match {
      case None => if (errorMessage != null) assert(false, errorMessage)
      case Some(resolver) => {
        val ctxValue = resolver.getPropertyValue(instance, prop)
        prop.setValue(instance, prepareValueForInjection(ctxValue))
      }
    }
  }

}

class ScalarInjection(prop: InjectableProperty,
                      resolvers: List[PropertyResolver]) extends Injection(prop, resolvers) {
  def inject[T](instance: T, values: List[Any]) = {
    values.size match {
      case 0 => if (!prop.isOptional) {
        findAndSetContextualValue(instance, "Missing value for mandatory property " + prop.getName + " in " + instance)
      } else if (prop.isDelegate) {
        // if the property is the delegate than the resolver must provide the delegate reference
        findAndSetContextualValue(instance, "Property resolver must provide a delegate reference for property "
          + prop.getName + " in " + instance)
      } else {
        // try to find the value for the optional property in the context - do not fire an exception if not found
        findAndSetContextualValue(instance, null /* null means no exception is fired*/)
      }
      case 1 => prop.setValue(instance, prepareValueForInjection(values.head))
      case _ => assert(false, "To many values for property " + prop.getName + " in " + instance)
    }

    instance
  }
}

class ArrayInjection(prop: InjectableProperty,
                     resolvers: List[PropertyResolver]) extends Injection(prop, resolvers) {
  def inject[T](instance: T, values: List[Any]) = {
    val componentType = prop.getArrayOrCollectionComponentType
    assert(!componentType.isArray,
      String.format("Multi-dimensional array property %s is not supported. Only single-dimensional arrays are supported",
        prop.getName))

    val array = java.lang.reflect.Array.newInstance(componentType, values.size)
    prop.setValue(instance, array)

    values.foldLeft(0)((i, v) => {
      val valToInject = prepareValueForInjection(v)
      prop.setArrayElement(instance, i, valToInject)

      i + 1
    })

    instance
  }
}

class CollectionInjection(prop: InjectableProperty,
                          resolvers: List[PropertyResolver]) extends Injection(prop, resolvers) {

  def inject[T](instance: T, values: List[Any]) = {

    if (classOf[java.util.Set[_]].isAssignableFrom(prop.getType)) {
      injectCollection(classOf[java.util.HashSet[_]])
    } else if (classOf[java.util.List[_]].isAssignableFrom(prop.getType)) {
      injectCollection(classOf[java.util.ArrayList[_]])
    } else {
      injectCollection()
    }

    def injectCollection(defaultCollectionClass: Class[_] = null) {
      var currentCollection = prop.getValue(instance).asInstanceOf[java.util.Collection[Any]]

      if (currentCollection == null) {
        currentCollection = defaultCollectionClass.newInstance.asInstanceOf[java.util.Collection[Any]]
        prop.setValue(instance, currentCollection)
      }

      values.foreach(v => {
        val valToInject: Any = prepareValueForInjection(v)

        prop.addCollectionElement(instance, valToInject)
      })
    }

    instance
  }
}

class MapInjection(prop: InjectableProperty,
                   resolvers: List[PropertyResolver]) extends Injection(prop, resolvers) {

  def inject[T](instance: T, values: List[Any]) = {
    val pairs = values.asInstanceOf[List[(AnyRef, AnyRef)]]
    var map: Map[AnyRef, AnyRef] = Map.empty
    pairs.foreach(pair => {
      val key = pair._1
      assert(!map.contains(key), "To many values for key " + key + " in property " + prop.getName + " in " + instance)
      val v = pair._2
      map += (key -> v)
    })

    var currentMap = prop.getValue(instance).asInstanceOf[java.util.Map[AnyRef, AnyRef]]

    if (currentMap == null) {
      currentMap = new java.util.HashMap[AnyRef, AnyRef]()
      prop.setValue(instance, currentMap)
    }

    map.foreach(p => {
      val k = prepareValueForInjection(p._1)
      val v = prepareValueForInjection(p._2)

      prop.putMapEntry(instance, k.asInstanceOf[AnyRef], v.asInstanceOf[AnyRef])
    })

    instance
  }
}
